### ISSUE TITLE
sbyte/short to nuint conversions are implicit

### ISSUE THREAD

**Issue by Naine** [dotnet/roslyn#49596]
**Version Used**: 3.8.0-5.20567.16 (53c5d7d3cf13d88978744a32a27c5f8350a8400a), Visual Studio 16.8.2

The spec at https://github.com/dotnet/csharplang/blob/master/proposals/csharp-9.0/native-integers.md says:

> | Operand | Target | Conversion | IL |
> |:---:|:---:|:---:|:---:|
> | `sbyte` | `nuint` | ExplicitNumeric | `conv.u` / `conv.ovf.u` |
> | `short` | `nuint` | ExplicitNumeric | `conv.u` / `conv.ovf.u` |

But the compiler allows these two conversions implicitly, which both change the value of negative input. If the casts are made explicit, IDE0004 is reported.

```csharp
static class NativeInts
{
    // These compile fine
    static nuint Implicit(sbyte x) => x;
    static nuint Implicit(short x) => x;

    // IDE0004 reported on both of these
    static nuint Explicit(sbyte x) => (nuint)x;
    static nuint Explicit(short x) => (nuint)x;
}
```

In a checked context, IDE0004 is still reported for explicit casts, and applying the code fixer affects the codegen.

```csharp
static class NativeInts
{
    static nuint CheckedExplicit(short x) => checked((nuint)x); // emits conv.ovf.u
    static nuint CheckedImplicit(short x) => checked(x);        // emits conv.i
}
```

I suspect that since C# 9 has shipped, this can no longer be fixed to match the current spec for compatibility reasons. That makes these the only built-in implicit conversions that can overflow. I think that at least IDE0004 should be updated to not suggest or offer to remove the "redundant" cast in a checked context.

**Comment by jaredpar** [comment]
@cston FYI

**Comment by cston** [comment]
@Naine, thanks for reporting this issue!
